# Contents List

## 01. Importing Libraries
## 02. Paths
## 03. Importing Data
## 04. Data Wrangling
## 05. Data Dictionary & Data Wrangling
## 06. Data Wrangling (Task 4.4)

# 01. Importing libraries

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Paths

In [5]:
#Creating a path varaible - Shortcut for Importing Data Files
path = r'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

In [7]:
path

'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

# 03. Importing Data

In [10]:
#Import the data the Orders Dataset
df_ords = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'orders.csv'), index_col = False)

In [11]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [14]:
#Import the data the Orders Dataset
df_prods = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'products.csv'), index_col = False)

In [16]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# 04. Data Wrangling

## 4.1 Dataset: Orders

In [20]:
# dropping the column "eval_set" & updating the datafarme
df_ords = df_ords.drop(columns = ['eval_set'])

In [22]:
df_ords

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [24]:
#Renaming the order_dow varaible to orders_day_of_week
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [26]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [28]:
# Changing data type of order_id variable to string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [30]:
df_ords['order_id'].dtype

dtype('O')

In [32]:
#Transposing Data

##Working with departments data
#1. Import the data 
df_dep = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'departments.csv'), index_col = False)

In [34]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [36]:
#Transposing the dataframe (df_dep)
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [38]:
#Assigning the transposed dataframe to a new dataframe (df_dep)
df_dep_t = df_dep.T

In [40]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [42]:
#Adding an Index to the dataframe (df_dep)
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [44]:
#Make the first row of the datafarme to header of the dataframe (df_dep)
new_header = df_dep_t.iloc[0]

In [46]:
new_header

0    department
Name: department_id, dtype: object

In [48]:
#Create a new dataframe that only copies over rows beyond the first row "the header row" (df_dep)
df_dep_t_new = df_dep_t[1:]

In [50]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [52]:
#Set the new_header variable as the new header in the dataframe "df_dep_t_new"
df_dep_t_new.columns = new_header

In [54]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 05. Data Dictionary & Data Wrangling

In [57]:
#Turn the datafarme to data dictionary "df_dep_t_new"
data_dict = df_dep_t_new.to_dict('index')

In [59]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [61]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [63]:
# Use the data dictionary to determine what then nummbers in department_id varaible stands for
print(data_dict.get('19'))

{'department': 'snacks'}


In [65]:
#Indexing, looking for element from a dataframe 
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [67]:
#Creating a new datafarme where the there is only department_id 19 "snacks"
df_snacks =  df_prods[df_prods['department_id']==19]

In [69]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [71]:
# Second version - Another way to find where the department_id is 19
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [73]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [75]:
# Third version - Another way to find where the department_id varaible is 19 and other ids within the department_id varaible
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [77]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# 06. Data Wrangling (Task 4.4)

### 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [81]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [83]:
# Changing data type of user_id variable to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [84]:
df_ords['user_id'].dtype

dtype('O')

### 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [88]:
#Renaming the days_since_prior_order varaible to days_since_last_order without updating the dataframe
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = False)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_last_order  
0                          NaN  
1                         15.0  
2                         21.0  
3                         29.0  
4                         28.0  
...                        ...  
3421078                   29.0  
3421079                   30.0  
3421080                   18.0  
3421081                    7.0  
3421082                   30.0  

[3421083 rows x 6 columns]

In [90]:
#Checking the update above has not been implemented in the dataframe
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [93]:
# The count/ frequency of the order_hour_of_day variable where all the column’s values are shown
df_ords['order_hour_of_day'].value_counts()

##Answer: The busiest hour for placing orders is 10 in the morning.

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

### 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [96]:
# Use the data dictionary to determine what nummber 4 in department_id varaible stands for
print(data_dict.get('4'))

## Answer: department_id = 4 stands for the produce department.

{'department': 'produce'}


### 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [99]:
#First find out the department_id for breakfast item
data_dict
#Answer: the breakfast department_id is 14

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [101]:
#Then we want to look closer at the sales of items in breakfast department
df_prods[df_prods['department_id']==19]

#Answer: A list of prducts names in the the breakfast department along with thier prices  

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

### 7. Details about products that customers might use to throw dinner parties. Find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood.


In [104]:
# Fidning all products from multiple department_ids
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5, 7, 20, 12])]

In [106]:
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?
##### The last dataframe I created was the df_dinner_parties which contains a total of 7650 rows.

### 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [110]:
#Locating customer with user_id = 1 (df_customer_id_one)
df_ords[df_ords['user_id']== '1']

#Answer: This customer placed a total of 11 orders. 
#One of their orders (order_id 2295261) shows 0.0 for the days_since_prior_order, which is most likely a mistak.

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [113]:
# To get more infomation about the customer with user_id = 1 - Some statistics 
df_ords[df_ords['user_id']=='1'].describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [119]:

df_dep_t_new.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'departments.csv'))

# 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [150]:
#Exporting / saving the wrangled df_ords to Prepared-Data
df_ords.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'orders_wrangled.csv'))

# 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in the “Prepared Data”

In [152]:
#Exporting / saving the wrangled df_ords to Prepared-Data
df_dep_t_new.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'departments_wrangled.csv'))